File: backend/ml/evaluate_model.py
"""Model evaluation helpers — mirrors ml/evaluate_model.ts."""

from __future__ import annotations

from typing import Any

from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix,
    precision_recall_fscore_support,
)


def evaluate_model(y_true, y_pred, class_names: list[str]) -> dict[str, Any]:
    """Compute accuracy, macro precision/recall/F1 and the confusion matrix."""
    precision, recall, f1, support = precision_recall_fscore_support(
        y_true, y_pred, labels=list(range(len(class_names))), average=None,
    )
    macro = precision_recall_fscore_support(
        y_true, y_pred, labels=list(range(len(class_names))), average="macro",
    )
    report = {
        "accuracy": float(accuracy_score(y_true, y_pred)),
        "precision": float(macro[0]),
        "recall": float(macro[1]),
        "f1": float(macro[2]),
        "confusion_matrix": confusion_matrix(
            y_true, y_pred, labels=list(range(len(class_names))),
        ).tolist(),
        "per_class": [
            {
                "disease": name,
                "precision": float(precision[i]),
                "recall": float(recall[i]),
                "f1": float(f1[i]),
                "support": int(support[i]),
            }
            for i, name in enumerate(class_names)
        ],
        "classification_report": classification_report(
            y_true, y_pred, labels=list(range(len(class_names))),
            target_names=class_names, zero_division=0,
        ),
    }
    return report

File: ml/evaluate_model.ts
/**
 * Model evaluation — mirrors backend/ml/evaluate_model.py.
 * Computes accuracy, macro-averaged precision/recall/F1 and the confusion
 * matrix over a held-out test set.
 */

import type { PerClassMetrics } from "../src/convex/ml/types";

export function confusionMatrix(yTrue: number[], yPred: number[], nClasses: number): number[][] {
  const matrix = Array.from({ length: nClasses }, () => new Array<number>(nClasses).fill(0));
  for (let i = 0; i < yTrue.length; i++) matrix[yTrue[i]][yPred[i]]++;
  return matrix;
}

export interface EvalResult {
  accuracy: number;
  precision: number;
  recall: number;
  f1: number;
  confusionMatrix: number[][];
  perClass: PerClassMetrics[];
  testSamples: number;
}

export function evaluate(
  yTrue: number[],
  yPred: number[],
  classes: string[],
): EvalResult {
  const nClasses = classes.length;
  const matrix = confusionMatrix(yTrue, yPred, nClasses);

  let correct = 0;
  for (let i = 0; i < yTrue.length; i++) if (yTrue[i] === yPred[i]) correct++;
  const accuracy = yTrue.length > 0 ? correct / yTrue.length : 0;

  const perClass: PerClassMetrics[] = [];
  let precisionSum = 0;
  let recallSum = 0;
  let f1Sum = 0;
  for (let c = 0; c < nClasses; c++) {
    const tp = matrix[c][c];
    let fp = 0;
    let fn = 0;
    let support = 0;
    for (let r = 0; r < nClasses; r++) {
      fp += matrix[r][c];
      fn += matrix[c][r];
      support += matrix[c][r];
    }
    fp -= tp;
    fn -= tp;
    const precision = tp + fp > 0 ? tp / (tp + fp) : 0;
    const recall = tp + fn > 0 ? tp / (tp + fn) : 0;
    const f1 = precision + recall > 0 ? (2 * precision * recall) / (precision + recall) : 0;
    precisionSum += precision;
    recallSum += recall;
    f1Sum += f1;
    perClass.push({ disease: classes[c], precision, recall, f1, support });
  }

  const macroPrecision = precisionSum / nClasses;
  const macroRecall = recallSum / nClasses;
  const macroF1 = f1Sum / nClasses;

  return {
    accuracy,
    precision: macroPrecision,
    recall: macroRecall,
    f1: macroF1,
    confusionMatrix: matrix,
    perClass,
    testSamples: yTrue.length,
  };
}
